### PR TITLE
#9486: Merge CCL reduce_scatter to TTNN

### PR DESCRIPTION
**Pull request by Aswinmcw**
### Ticket
#9486 

### Problem description
CCL ops needs to be merged to TTNN

### What's changed
Merge CCL reduce scatter to TTNN

### Checklist
- [x] All Post Commit : https://github.com/tenstorrent/tt-metal/actions/runs/9934253961
- [x] T3000 Frequent Test : https://github.com/tenstorrent/tt-metal/actions/runs/9956468935
- [x] T3000 Profiler Test : https://github.com/tenstorrent/tt-metal/actions/runs/9920010771
- [x] T3000 Unit Test : https://github.com/tenstorrent/tt-metal/actions/runs/9871804792
- [x] T3000 Demo Test : https://github.com/tenstorrent/tt-metal/actions/runs/9956476040

### CODE REVIEW

**Review by SeanNijjar — COMMENTED** [review]
Hey @Aswinmcw, this is looking good although I think there are a couple things missing in this PR still, namely the updates to the test invocations in `tests/scripts/run_tests.sh` and `tests/scripts/t3000/run_t3000_frequent_tests.sh`

You'll see calls to the `test_reduce_scatter` scripts and there and the paths will need to be updated.

Likewise, the gtest invocations in `run_tt_eager.py` will need to be updated to point to the new paths.

I saw a `run_ttnn.sh` test file but looking at the contents, I'm not sure if it's appropriate to move the test invocations there. I'll ask around and get back to you about this. For now the paths should be updated at the least.

**Comment by SeanNijjar** [comment]
> I saw a `run_ttnn.sh` test file but looking at the contents, I'm not sure if it's appropriate to move the test invocations there. I'll ask around and get back to you about this. For now the paths should be updated at the least.

Let's just do the path updates for now for the gtests - no need to move the test invocations to a new file as that'll fundamentally change where/how the tests are run.

**Comment by Aswinmcw** [comment]
@SeanNijjar i have just updated the test paths for gtest and moved the pytest to TTNN. Please review it once

**Comment by Aswinmcw** [comment]
> Hey @Aswinmcw, this is looking good although I think there are a couple things missing in this PR still, namely the updates to the test invocations in `tests/scripts/run_tests.sh` and `tests/scripts/t3000/run_t3000_frequent_tests.sh`
> 
> You'll see calls to the `test_reduce_scatter` scripts and there and the paths will need to be updated.
> 
> Likewise, the gtest invocations in `run_tt_eager.py` will need to be updated to point to the new paths.
> 
> I saw a `run_ttnn.sh` test file but looking at the contents, I'm not sure if it's appropriate to move the test invocations there. I'll ask around and get back to you about this. For now the paths should be updated at the least.

`test_reduce_scatter` is called only in `tests/scripts/t3000/run_t3000_frequent_tests.sh` , updated the path there.

**Comment by SeanNijjar** [comment]
> > Hey @Aswinmcw, this is looking good although I think there are a couple things missing in this PR still, namely the updates to the test invocations in `tests/scripts/run_tests.sh` and `tests/scripts/t3000/run_t3000_frequent_tests.sh`
> > You'll see calls to the `test_reduce_scatter` scripts and there and the paths will need to be updated.
> > Likewise, the gtest invocations in `run_tt_eager.py` will need to be updated to point to the new paths.
> > I saw a `run_ttnn.sh` test file but looking at the contents, I'm not sure if it's appropriate to move the test invocations there. I'll ask around and get back to you about this. For now the paths should be updated at the least.
> 
> `test_reduce_scatter` is called only in `tests/scripts/t3000/run_t3000_frequent_tests.sh` , updated the path there.

There is also a `test_reduce_scatter_nightly` in  run_tests.sh

**Comment by Aswinmcw** [comment]
> > > Hey @Aswinmcw, this is looking good although I think there are a couple things missing in this PR still, namely the updates to the test invocations in `tests/scripts/run_tests.sh` and `tests/scripts/t3000/run_t3000_frequent_tests.sh`
> > > You'll see calls to the `test_reduce_scatter` scripts and there and the paths will need to be updated.
> > > Likewise, the gtest invocations in `run_tt_eager.py` will need to be updated to point to the new paths.
> > > I saw a `run_ttnn.sh` test file but looking at the contents, I'm not sure if it's appropriate to move the test invocations there. I'll ask around and get back to you about this. For now the paths should be updated at the least.
> > 
> > 
> > `test_reduce_scatter` is called only in `tests/scripts/t3000/run_t3000_frequent_tests.sh` , updated the path there.
> 
> There is also a `test_reduce_scatter_nightly` in run_tests.sh

I see the `test_reduce_scatter_nightly` function only in the test file `tests/tt_eager/python_api_testing/unit_testing/misc/test_reduce_scatter_nightly.py` . Am i missing something?
Screenshot 2024-07-05 at 8 26 36 PM

**Comment by SeanNijjar** [comment]
> > > > Hey @Aswinmcw, this is looking good although I think there are a couple things missing in this PR still, namely the updates to the test invocations in `tests/scripts/run_tests.sh` and `tests/scripts/t3000/run_t3000_frequent_tests.sh`
> > > > You'll see calls to the `test_reduce_scatter` scripts and there and the paths will need to be updated.
> > > > Likewise, the gtest invocations in `run_tt_eager.py` will need to be updated to point to the new paths.
> > > > I saw a `run_ttnn.sh` test file but looking at the contents, I'm not sure if it's appropriate to move the test invocations there. I'll ask around and get back to you about this. For now the paths should be updated at the least.
> > > 
> > > 
> > > `test_reduce_scatter` is called only in `tests/scripts/t3000/run_t3000_frequent_tests.sh` , updated the path there.
> > 
> > 
> > There is also a `test_reduce_scatter_nightly` in run_tests.sh
> 
> I see the `test_reduce_scatter_nightly` function only in the test file `tests/tt_eager/python_api_testing/unit_testing/misc/test_reduce_scatter_nightly.py` . Am i missing something?

Maybe I am missing something but `test_reduce_scatter_nightly.py` is currently run from `run_tests.sh` but I don't see that file updated to point to the new location of `test_reduce_scatter_nightly.py`

**Comment by Aswinmcw** [comment]
> > > > > Hey @Aswinmcw, this is looking good although I think there are a couple things missing in this PR still, namely the updates to the test invocations in `tests/scripts/run_tests.sh` and `tests/scripts/t3000/run_t3000_frequent_tests.sh`
> > > > > You'll see calls to the `test_reduce_scatter` scripts and there and the paths will need to be updated.
> > > > > Likewise, the gtest invocations in `run_tt_eager.py` will need to be updated to point to the new paths.
> > > > > I saw a `run_ttnn.sh` test file but looking at the contents, I'm not sure if it's appropriate to move the test invocations there. I'll ask around and get back to you about this. For now the paths should be updated at the least.
> > > > 
> > > > 
> > > > `test_reduce_scatter` is called only in `tests/scripts/t3000/run_t3000_frequent_tests.sh` , updated the path there.
> > > 
> > > 
> > > There is also a `test_reduce_scatter_nightly` in run_tests.sh
> > 
> > 
> > I see the `test_reduce_scatter_nightly` function only in the test file `tests/tt_eager/python_api_testing/unit_testing/misc/test_reduce_scatter_nightly.py` . Am i missing something?
> 
> Maybe I am missing something but `test_reduce_scatter_nightly.py` is currently run from `run_tests.sh` but I don't see that file updated to point to the new location of `test_reduce_scatter_nightly.py`

Or can we add the reduce scatter test to the `run_tests.sh` as like all gather test?

**Comment by SeanNijjar** [comment]
> Or can we add the reduce scatter test to the `run_tests.sh` as like all gather test?

Ah sorry for the confusion, I thought this change was mainlined but I guess it wasn't, it's on branch. Please disregard my prior comments.

**Review by TT-BrianLiu — APPROVED** [review]

**Comment by SeanNijjar** [comment]
Approved pending pipeline completion:
https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml
https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-frequent-tests.yaml

File: tests/ttnn/unit_tests/operations/test_distributed_layernorm.py
Comment (eyonland) [inline tests/ttnn/unit_tests/operations/test_distributed_layernorm.py:43]: @SeanNijjar is this the name we want?  Our objective is to make our API familiar to PyTorch users.

Comment (SeanNijjar) [reply tests/ttnn/unit_tests/operations/test_distributed_layernorm.py:43]: Hey @eyonland, which part of the expression are you referring to here? If the RHS, I am not sure. I'm not up to speed on what's idiomatic pytorch so I couldn't say if the name is appropriate. 

In c++ an aggregate is more like a struct so It doesn't _sound_ like what I'd expect at first but it looks like python frameworks like pandas use aggregate for collecting some dataset so it could apply. 

I don't have a valuable opinion on this.

**Review by tt-aho — APPROVED** [review]
lgtm

File: tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py:80]: I assume @cfjchu will have the same request here to use the new mesh tensor infrastructure. I view it as somewhat orthogonal to moving the ops to TTNN so I personally would not hold up the PR and would just make sure to do that  after the move is done.

**Comment by SeanNijjar** [comment]
@Aswinmcw please run the same pipelines as were run for the line_all_gather migration, including the t3000 model tests.

Thank you.

File: tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py
Comment (cfjchu) [reply tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py:80]: Yes, as long as it is not dropped. The primary way that users should be using the op should be via multi-device tensors so I'd expect this change to be done so it's usable.

Comment (cfjchu) [reply tests/ttnn/unit_tests/operations/test_reduce_scatter_post_commit.py:80]: I'm trying this here: https://github.com/tenstorrent/tt-metal/issues/10296

**Review by SeanNijjar — APPROVED** [review]
Approved pending clean pipelines (or no regressions for those pipelines that aren't green on main)

File: ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/host/reduce_scatter_full_worker_grid.cpp:856]: unclear why this is in ttnn::utils.. and not ttnn::ccl or ttnn::ccl::utils

**Review by ayerofieiev-tt — APPROVED** [review]
Please review my last comment on the utils namespace

**Review by ttmchiou — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by johanna-rock-tt — APPROVED** [review]
Falcon40b changes look good to me

**Pull request merged**
